File: golang/_0031_Next_Permutation/main.go
package main

func nextPermutation(nums []int) {
	n := len(nums)
	if n <= 1 {
		return
	}

	i, j := n-1, n-1
	for i > 0 && nums[i] <= nums[i-1] {
		i--
	}

	if i == 0 {
		reverse(nums)
	} else {
		for nums[j] <= nums[i-1] {
			j--
		}
		swap(nums, i-1, j)
		reverse(nums[i:])
	}
}

func reverse(nums []int) {
	for i, j := 0, len(nums)-1; i < j; i, j = i+1, j-1 {
		swap(nums, i, j)
	}
}

func swap(nums []int, i, j int) {
	nums[i], nums[j] = nums[j], nums[i]
}

File: python3/_0031_Next_Permutation/main.py
from typing import List


class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        if not nums:
            return

        n = len(nums)
        j = n - 1
        while j > 0:
            if nums[j] > nums[j - 1]:
                break
            j -= 1

        if j == 0:
            nums.reverse()
        else:
            i = j + 1
            while i < n and nums[i] > nums[j - 1]:
                i += 1

            nums[j - 1], nums[i - 1] = nums[i - 1], nums[j - 1]

            u, v = j, n - 1
            while u < v:
                nums[u], nums[v] = nums[v], nums[u]
                u += 1
                v -= 1
